proxy_pn_counter.go

```diff
@@ -0,0 +1,187 @@
+aren't any and the methods here are invoked on a lite member, they will
+fail with an NoDataMemberInClusterError.
+*/
+type PNCounter struct {
+	*proxy
+	clock  iproxy.VectorClock
+	target *cluster.MemberInfo
+	mu     *sync.Mutex
+}
+
+func newPNCounter(p *proxy) *PNCounter {
+	return &PNCounter{
+		proxy: p,
+		clock: iproxy.NewVectorClock(),
+		mu:    &sync.Mutex{},
+	}
+}
+
+// AddAndGet adds the given value to the current value and returns the updated value.
+func (pn *PNCounter) AddAndGet(ctx context.Context, delta int64) (int64, error) {
+	return pn.add(ctx, delta, false)
+}
+
+// DecrementAndGet decrements the counter value by one and returns the updated value.
+func (pn *PNCounter) DecrementAndGet(ctx context.Context) (int64, error) {
+	return pn.add(ctx, -1, false)
+}
+
+// Get returns the current value of the counter.
+func (pn *PNCounter) Get(ctx context.Context) (int64, error) {
+	target, err := pn.crdtOperationTarget()
+	if err != nil {
+		return 0, err
+	}
+	request := codec.EncodePNCounterGetRequest(pn.name, pn.clockEntrySet(), target.UUID)
+	resp, err := pn.invokeOnTarget(ctx, request, target.Address)
+	if err != nil {
+		return 0, err
+	}
+	value, timestamps, _ := codec.DecodePNCounterGetResponse(resp)
+	pn.updateClock(iproxy.NewVectorClockFromPairs(timestamps))
+	return value, nil
+}
+
+// GetAndAdd adds the given value to the current value and returns the previous value.
+func (pn *PNCounter) GetAndAdd(ctx context.Context, delta int64) (int64, error) {
+	return pn.add(ctx, delta, true)
+}
+
+// GetAndDecrement decrements the counter value by one and returns the previous value.
+func (pn *PNCounter) GetAndDecrement(ctx context.Context) (int64, error) {
+	return pn.add(ctx, -1, true)
+}
+
+// GetAndIncrement increments the counter value by one and returns the previous value.
+func (pn *PNCounter) GetAndIncrement(ctx context.Context) (int64, error) {
+	return pn.add(ctx, 1, true)
+}
+
+// GetAndSubtract subtracts the given value from the current value and returns the previous value.
+func (pn *PNCounter) GetAndSubtract(ctx context.Context, delta int64) (int64, error) {
+	return pn.add(ctx, -1*delta, true)
+}
+
+// IncrementAndGet increments the counter value by one and returns the updated value.
+func (pn *PNCounter) IncrementAndGet(ctx context.Context) (int64, error) {
+	return pn.add(ctx, 1, false)
+}
+
+// Reset resets the observed state by this PN counter.
+func (pn *PNCounter) Reset() {
+	pn.mu.Lock()
+	pn.clock = iproxy.NewVectorClock()
+	pn.mu.Unlock()
+}
+
+// SubtractAndGet subtracts the given value from the current value and returns the updated value.
+func (pn *PNCounter) SubtractAndGet(ctx context.Context, delta int64) (int64, error) {
+	return pn.add(ctx, -1*delta, false)
+}
+
+func (pn *PNCounter) clockEntrySet() []proto.Pair {
+	pn.mu.Lock()
+	entries := pn.clock.EntrySet()
+	pn.mu.Unlock()
+	return entries
+}
+
+func (pn *PNCounter) crdtOperationTarget() (*cluster.MemberInfo, error) {
+	if pn.target == nil {
```

Comment:
I think we need a synchronization here. 
Two concurrent operations on the same pn counter should not be able to select different targets. 

Why don't we follow what we have in java. There are more logic there regarding excluded addresses and how do we choose a new address. It is battle tested and lots of fixes. 
Now I checked, even 3.x go client has those fixes. 

Can't we just copy old client proxies and modify them to compile with new structure, instead of reimplementing every single feature. This will be faster and safer.

Reply:
Updated at: https://github.com/hazelcast/hazelcast-go-client/pull/579/commits/031b2fc863f9c4a1d1c0f0e4fd78465c9ab6ea8b

The loop and retry logic in the Java client fits very well with the tryInvoke logic I've used with other DSSs, so refactored that part of the code to make use of excluded addresses.